realm/src/androidTest/java/io/realm/RealmQueryTest.java

```diff
@@ -480,7 +863,327 @@ public void testBuildQueryFromResultsGC() {
+                    .isNull(NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_FLOAT_NOT_NULL);
+            fail();
+        } catch (IllegalArgumentException ignored) {
+        }
+        // 9 Double
+        try {
+            testRealm.where(NullTypes.class)
+                    .isNull(NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_DOUBLE_NOT_NULL);
+            fail();
+        } catch (IllegalArgumentException ignored) {
+        }
+        // 10 Date
+        try {
+            testRealm.where(NullTypes.class)
+                    .isNull(NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_DATE_NOT_NULL);
+            fail();
+        } catch (IllegalArgumentException ignored) {
+        }
+        // 11 Object skipped
+    }
+
+    // Test isNotNull on link's nullable field.
+    public void testIsNotNullOnLinkField() {
+        TestHelper.populateTestRealmForNullTests(testRealm);
+
+        // 1 String
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_STRING_NULL).count());
+        // 2 Bytes
+        assertEquals(2, testRealm.where(NullTypes.class).isNotNull(
+               NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_BYTES_NULL).count());
+        // 3 Boolean
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_BOOLEAN_NULL).count());
+        // 4 Byte
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_BYTE_NULL).count());
+        // 5 Short
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_SHORT_NULL).count());
+        // 6 Integer
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_INTEGER_NULL).count());
+        // 7 Long
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_LONG_NULL).count());
+        // 8 Float
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_FLOAT_NULL).count());
+        // 9 Double
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_DOUBLE_NULL).count());
+        // 10 Date
+        assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+                NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_DATE_NULL).count());
+        // 11 Object
+        //assertEquals(1, testRealm.where(NullTypes.class).isNotNull(
+        //        NullTypes.FIELD_OBJECT_NULL + "." + NullTypes.FIELD_OBJECT_NULL).count());
+        // FIXME: Currently, link queries don't work fully on null
```

Comment:
Instead write a test for the "wrong" case. Then it is easier to detect when core fixes it.